Add Matmul recipe into x86_inductor_quantizer

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123923
* #122777
* __->__ #122776

**Summary**
Add `matmul` in the quantization recipes, noting that it's not a general recipe but tailored to meet accuracy criteria for specific models. `matmul` recipe is disabled by default.

**Test Plan**
```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```

Differential Revision: [D56288468](https://our.internmc.facebook.com/intern/diff/D56288468)

leslie-fang-intel committed:
Add Matmul recipe into x86_inductor_quantizer

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/quantizer/x86_inductor_quantizer.py:
```diff
@@ -82,7 +82,9 @@ class _X86InductorQuantizationAnnotation(QuantizationAnnotation):
 # A superset of default_quantizable_ops includes operators support the int8 data type
 # but not enabled by default recipe of X86InductorQuantizer.
-quantizable_ops = default_quantizable_ops
+quantizable_ops = default_quantizable_ops | {
```
why not just add matmul to default_quantizable_ops?

leslie-fang-intel replied:
Because we think it's not a general recipe. `matmul` recipe has been used for Bert-Large and VIT, but not used for most of the LLM models because of the accuracy concern. So, we decide don't put it into the `default_quantizable_ops`.

jerryzh168 replied:
I mean what is the difference, since you are annotating it in the quantizer anyways?

leslie-fang-intel replied:
Oh, got it. Since `matmul` is not in `default_quantizable_ops`, `if config := self._get_aten_operator_type(torch.ops.aten.matmul.default):` will return None by default and we will not annotate `matmul` actually. Unless, we use `_set_aten_operator_type` to set the Quantization config for `matmul`.

jerryzh168 replied:
OK, I see. sounds good

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
Hi @leslie-fang-intel! Sorry, I need to revert your PR because it breaks the internal build. Here is the error message:
```
_internal/common_quantization.py", line 769, in checkGraphModuleNodes
    self.assertTrue(
AssertionError: False is not true : Check failed for graph:'get_attr' '_scale_0' '_scale_0' () {}
'get_attr' '_zero_point_0' '_zero_point_0' () {}
'get_attr' '_frozen_param0' '_frozen_param0' () {}
'call_function' 'dequantize_per_channel_default' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param0, _scale_0, _zero_point_0, 0, -128, 127, torch.int8) {}
'get_attr' '_scale_1' '_scale_1' () {}
'get_attr' '_zero_point_1' '_zero_point_1' () {}
'get_attr' '_frozen_param1' '_frozen_param1' () {}
'call_function' 'dequantize_per_channel_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param1, _scale_1, _zero_point_1, 0, -128, 127, torch.int8) {}
'get_attr' '_scale_2' '_scale_2' () {}
'get_attr' '_zero_point_2' '_zero_point_2' () {}
'get_attr' '_frozen_param2' '_frozen_param2' () {}
'call_function' 'dequantize_per_channel_default_2' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')> (_frozen_param2, _scale_2, _zero_point_2, 0, -128, 127, torch.int8) {}
'placeholder' 'x' 'x' () {}
'call_function' 'quantize_per_tensor_default' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')> (x, 0.03170151636004448, 128, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default_3' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.03170151636004448, 128, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default_2' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.03170151636004448, 128, 0, 255, torch.uint8) {}
'call_function' 'dequantize_per_tensor_default_1' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')> (quantize_per_tensor_default, 0.03170151636004448, 128, 0, 255, torch.uint8) {}
'call_function' 'linear' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_1, dequantize_per_channel_default) {}
'call_function' 'linear_1' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_2, dequantize_per_channel_default_1) {}
'call_function' 'linear_2' <OpOverload(op='aten.linear', overload='default')> (dequantize_per_tensor_default_3, dequantize_per_channel_default_2) {}
'call_function' 'view' <OpOverload(op='aten.view', overload='default')> (linear, [2, 384, 16, 64]) {}
'call_function' 'permute' <OpOverload(op='aten.permute', overload='default')> (view, [0, 2, 1, 3]) {}
'call_function' 'view_1' <OpOverload(op='aten.view', overload='default')> (linear_1, [2, 384, 16, 64]) {}
'call_function' 'permute_1' <OpOverload(op='aten.permute', overload='default')> (view_1, [0, 2, 1, 3]) {}
'call_function' 'view_2' <OpOverload(op='aten.view', overload='default')> (linear_2, [2, 384, 16, 64]) {}
'call_function' 'permute_2' <OpOverload(op='aten.permute', overload='default')> (view_2, [0, 2, 1, 3]) {}
'call_function' 'transpose' <OpOverload(op='aten.transpose', overload='int')> (permute_1, -1, -2) {}
'call_function' 'matmul' <OpOverload(op='aten.matmul', overload='default')> (permute, transpose) {}
'call_function' 'div' <OpOverload(op='aten.div', overload='Tensor')> (matmul, 32.0) {}
'call_function' 'softmax' <OpOverload(op='aten.softmax', overload='int')> (div, -1) {}
'call_function' 'matmul_1' <OpOverload(op='aten.matmul', overload='default')> (softmax, permute_2) {}
'output' 'output_1' 'output' ((matmul_1,),) {}Expected ordered list:['call_function' <OpOverload(op='quantized_decomposed.quantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_tensor', overload='default')>, 'call_function' <OpOverload(op='quantized_decomposed.dequantize_per_channel', overload='default')>, 'call_function' <OpOverload(op='aten.linear', overload='default')>, 'call_function' <OpOverload(op='aten.view', overload='default')>, 'call_function' <OpOverload(op='aten.permute', overload='default')>, 'call_function' <OpOverload(op='aten.matmul', overload='default')>, 'call_function' <OpOverload(op='aten.div', overload='Tensor')>, 'call_function' <OpOverload(op='aten.softmax', overload='int')>]
```
This is already the second such problem. We created a separate issue to discuss this - https://github.com/pytorch/pytorch/issues/123865

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
Thanks @DanilBaibak for sharing the failure msg. Let's discuss in https://github.com/pytorch/pytorch/issues/123865

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
Fixed the UT randomly failure. @DanilBaibak @jerryzh168 could you help to import the PR and check if any further internal CI failure?

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @clee2000 @DanilBaibak @jerryzh168, could you help to import this PR and check the internal preCI status?

jerryzh168 commented:
@pytorchbot rebase

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
Manually rebased by fixing conflict. @jerryzh168 please help to continue the import.

leslie-fang-intel committed:
Update on "Add Matmul recipe into x86_inductor_quantizer"


**Summary**
Add `matmul` in the quantization recipes, noting that it's not a general recipe but tailored to meet accuracy criteria for specific models. `matmul` recipe is disabled by default.

**Test Plan**
```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```


[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56288468).

leslie-fang-intel commented:
Hi @jerryzh168, thanks for importing the PR. May I know the internal testing result?

jerryzh168 commented:
@leslie-fang-intel internal CI is green, please feel free to merge

leslie-fang-intel commented:
> @leslie-fang-intel internal CI is green, please feel free to merge

Thanks, I will land it.

leslie-fang-intel commented:
@pytorchbot merge
